File: api-gateway-java-reactive/src/main/java/ru/kuramshindev/apigatewayjavareactive/UpstreamAsyncClients.java
package ru.kuramshindev.apigatewayjavareactive;


import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;
import ru.kuramshin.dto.companies.CompanyDto;
import ru.kuramshin.dto.contractors.ContractorDto;
import ru.kuramshin.dto.projects.ProjectDto;

import java.util.UUID;

@Slf4j
@RequiredArgsConstructor
@Component
public class UpstreamAsyncClients {

    private final WebClient webClient;

    public Mono<CompanyDto> getCompanyByIdAsync(UUID uuid) {
        return webClient.get().uri("/companies/{uuid}", uuid).retrieve().bodyToMono(CompanyDto.class);
    }

    public Mono<ContractorDto> getContractorByIdAsync(UUID uuid) {
        return webClient.get().uri("/contractors/{uuid}", uuid).retrieve().bodyToMono(ContractorDto.class);
    }

    public Mono<ProjectDto> getProjectByIdAsync(UUID uuid) {
        return webClient.get().uri("/projects/{uuid}", uuid).retrieve().bodyToMono(ProjectDto.class);
    }
}

File: api-gateway-kotlin-reactive/src/main/kotlin/ru/kuramshindev/apigatewaykotlinreactive/UpstreamAsyncClients.kt
package ru.kuramshindev.apigatewaykotlinreactive

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import ru.kuramshin.dto.companies.CompanyDto
import ru.kuramshin.dto.contractors.ContractorDto
import ru.kuramshin.dto.projects.ProjectDto
import java.util.*

@Component
class UpstreamAsyncClients(
    private val webClient: WebClient
) {
    suspend fun getCompanyByIdAsync(uuid: UUID): CompanyDto =
        webClient.get().uri("/companies/${uuid}").retrieve().awaitBody()

    suspend fun getContractorByIdAsync(uuid: UUID): ContractorDto =
        webClient.get().uri("/contractors/${uuid}").retrieve().awaitBody()

    suspend fun getProjectByIdAsync(uuid: UUID): ProjectDto =
        webClient.get().uri("/projects/${uuid}").retrieve().awaitBody()
}